### PR TITLE
Fix calling deprecate with correct arguments

### LINKED ISSUE #44702: async_status fails when job output contains depreciation warning
##### SUMMARY
I have been running several jobs (creating NICs), for some of them I couldn't obtain a status.

```
TASK [Get the status of the job] ***************************************************************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: deprecate requires a string not a <type 'dict'>
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):
  File \"/home/zim/.ansible/tmp/ansible-tmp-1535331789.16-166798681023581/AnsiballZ_async_status.py\", line 113, in <module>
    _ansiballz_main()
  File \"/home/zim/.ansible/tmp/ansible-tmp-1535331789.16-166798681023581/AnsiballZ_async_status.py\", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File \"/home/zim/.ansible/tmp/ansible-tmp-1535331789.16-166798681023581/AnsiballZ_async_status.py\", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File \"/tmp/ansible_async_status_payload_LzN9Fd/__main__.py\", line 101, in <module>
  File \"/tmp/ansible_async_status_payload_LzN9Fd/__main__.py\", line 97, in main
  File \"/tmp/ansible_async_status_payload_LzN9Fd/ansible_async_status_payload.zip/ansible/module_utils/basic.py\", line 2339, in exit_json
  File \"/tmp/ansible_async_status_payload_LzN9Fd/ansible_async_status_payload.zip/ansible/module_utils/basic.py\", line 2325, in _return_formatted
  File \"/tmp/ansible_async_status_payload_LzN9Fd/ansible_async_status_payload.zip/ansible/module_utils/basic.py\", line 961, in deprecate
TypeError: deprecate requires a string not a <type 'dict'>
",
    "module_stdout": "", 
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", 
    "rc": 1
}
        to retry, use: --limit @/mnt/c/dev-current/openshift-container-platform/ansible/playbooks/create.retry
```

The jobs that I couldn't obtain a status contained following element:
```
{
    "invocation": {
….
    "deprecations": [{
        "msg": "Setting ip_configuration flatten is deprecated and will be removed. Using ip_configurations list to define the ip configuration",
        "version": "2.9"
    }],
….
```

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
async_status

##### ANSIBLE VERSION
2.6
##### EXPECTED RESULTS
Job status should be returned
##### ACTUAL RESULTS
Parsing job output file fails

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
This fixes #44702

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
async_status (via basic.py)

##### ANSIBLE VERSION
v2.7

### CODE REVIEW

**Comment by dagwieers** [comment]
@zikalino Can you test this fix ?

File: lib/ansible/module_utils/basic.py [hunk 1]
```diff
@@ -2329,6 +2329,8 @@ def _return_formatted(self, kwargs):
                         self.deprecate(d[0], version=d[1])
                     else:
                         self.deprecate(d)
+            elif isinstance(kwargs['deprecations'], dict):
+                self.deprecate(kwargs['deprecations']['message'], version=kwargs['deprecations']['version'])
```
Comment (sivel) [inline lib/ansible/module_utils/basic.py]: I have not looked at this in depth, but I believe this should be `msg` instead of `message`.

Comment (dagwieers) [reply lib/ansible/module_utils/basic.py]: Correct, I was fixing that before I saw your comment :-)

**Review by achinthagunasekara — APPROVED** [review]

**Comment by zikalino** [comment]
@dagwieers thanks for fixing it so quickly! I will try it today and let you know!

**Comment by machacekondra** [comment]
We also hit this issue[1]. I've tried with this patch and it works fine! shipit

[1] https://bugzilla.redhat.com/show_bug.cgi?id=1630865

**Comment by mwperina** [comment]
Ryan, could this be merged and backported to Ansible 2.7 before GA?
@ryansb

File: lib/ansible/module_utils/basic.py [hunk 2]
```diff
@@ -2327,6 +2327,8 @@ def _return_formatted(self, kwargs):
                 for d in kwargs['deprecations']:
                     if isinstance(d, SEQUENCETYPE) and len(d) == 2:
                         self.deprecate(d[0], version=d[1])
+                    elif isinstance(d, Mapping):
+                        self.deprecate(d['msg'], version=d.get('version', None))
```
Comment (bcoca) [inline lib/ansible/module_utils/basic.py:2331]: what if msg does not exist?, shouldn't it also be a get

Comment (bcoca) [reply lib/ansible/module_utils/basic.py:2331]: in any case, deprecations should always be a list, unsure how a dict is making its way here

Comment (dagwieers) [reply lib/ansible/module_utils/basic.py:2331]: Both python and powershell module_utils define it as a list of dictionaries.
- https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L959
- https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/powershell/Ansible.ModuleUtils.Legacy.psm1#L125

It originates from this: https://github.com/ansible/ansible/pull/20884 and I reimplemented it for Windows.

Comment (bcoca) [reply lib/ansible/module_utils/basic.py:2331]: ah, nvmd, misread the code, thought you were operating on the 'top level' ... yes it is a list of dictionaries but we allowed for pre-existing string cases

Comment (bcoca) [reply lib/ansible/module_utils/basic.py:2331]: unsure why we need mapping, need to look at original bug, the self.deprecate(d) should be able to handle dicts already

Comment (dagwieers) [reply lib/ansible/module_utils/basic.py:2331]: So I guess this should be backported as well.

Comment (dagwieers) [reply lib/ansible/module_utils/basic.py:2331]: If the `module.deprecate()` call was used, the message cannot be missing. And we rewrote all modules to use this call.

Comment (bcoca) [reply lib/ansible/module_utils/basic.py:2331]: yeah, it makes sense now, i was just being dumb, modules doing right thing, async_wrapper is acting as proxy but not intelligently so it is passing the final artifact, which we have to reprocess again

**Comment by mwperina** [comment]
Brian, will this be included in 2.7.0?
@bcoca

**Comment by gundalow** [comment]
@mwperina Fix was released in Ansible 2.7.1

**Pull request merged**